### PR TITLE
[INTEL MKL] Enabled filter caching for convolution.

### PR DESCRIPTION
**Pull request by bhavani-subramanian**
When filter is a constant in convolution operations, we can avoid converting filter from TensorFlow format to MKL format by caching it after the first conversion. This cached filter can then be reused in subsequent iterations without having to perform the conversion again.

This PR also contains unit tests for the following cases:

1. To check if the attribute `is_filter_const` is set to true (in the graph pass) when filter is a constant and false otherwise.
2. To ensure that cached filter data is reused when `Conv2D` kernel is run more than once.

For quantized Conv2D kernels, we expect `is_filter_const` to be true in the compute kernel.

### CODE REVIEW

**Comment by bhavani-subramanian** [comment]
Hi @tatianashp, @penpornk can you please take a look at this PR? Thanks.

**Comment by penpornk** [comment]
@bhavani-subramanian Sorry for the delay! I'll look at it today.

**Review by penpornk — CHANGES_REQUESTED** [review]
Apologies again for the lateness! Thank you for the PR and for improving the code's readability!

File: tensorflow/core/kernels/mkl_fused_ops_test.cc [hunk 1]
```diff
@@ -401,5 +401,70 @@ TEST_F(FusedPadConvOpTest, PaddingConvTestNchw) {
...
+    AddInputFromArray<T>(image.shape(), image.flat<T>());
+    AddInputFromArray<T>(filter.shape(), filter.flat<T>());
+    AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+    AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+
+    TF_ASSERT_OK(RunOpKernel());
+
+    // Compare output to expected results
+    const Tensor& first = *GetOutput(0);
+    const Tensor& second = *GetOutput(2);
```
[the remark this answers is not in the document]
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_fused_ops_test.cc]: Done.

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 2]
```diff
@@ -1310,6 +1355,67 @@ class MklConvOp : public OpKernel {
 
     stream(stream::kind::eager).submit(net).wait();
   }
+
+  inline bool IsFilterCacheEmpty(OpKernelContext* context) {
+    tf_shared_lock lock(mu_);
+    SHARED_LOCKS_REQUIRED(mu_) {
```
[the remark this answers is not in the document]
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done.

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 3]
```diff
@@ -1310,6 +1355,67 @@ class MklConvOp : public OpKernel {
...
+  }
+
+  // Cache the converted filter in a persistent tensor.
+  // Only one thread can execute this method at any given time.
+  void CacheFilter(OpKernelContext* context,
+                   const std::shared_ptr<ConvFwdPd>& conv_fwd_pd,
+                   Tfilter* filter_data, const Tensor& filter_tensor,
+                   MklDnnData<Tfilter>& filter, const memory::desc& filter_md) {
+    mutex_lock lock(mu_);
+    EXCLUSIVE_LOCKS_REQUIRED(mu_) {
```
[the remark this answers is not in the document]
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done.

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 4]
```diff
@@ -1310,6 +1355,67 @@ class MklConvOp : public OpKernel {
...
+      return (cached_filter_data_tensor.NumElements() == 0);
+    }
+  }
+
+  // Cache the converted filter in a persistent tensor.
+  // Only one thread can execute this method at any given time.
+  void CacheFilter(OpKernelContext* context,
+                   const std::shared_ptr<ConvFwdPd>& conv_fwd_pd,
+                   Tfilter* filter_data, const Tensor& filter_tensor,
+                   MklDnnData<Tfilter>& filter, const memory::desc& filter_md) {
```
[the remark this answers is not in the document]
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: @penpornk I don't quite follow. Are you saying that I should include both `LOCKS_EXCLUDED` and `SHARED_LOCKS_REQUIRED'? Aren't they mutually exclusive?

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 5]
```diff
@@ -1310,6 +1355,67 @@ class MklConvOp : public OpKernel {
...
+      size_t cached_filter_data_size =
+          filter.GetOpMem().get_primitive_desc().get_size();
+      memcpy(cached_filter_data, filter_data, cached_filter_data_size);
+    }
+  }
+
+  Tfilter* GetCachedFilter(OpKernelContext* context,
+                           const memory::format& filter_mf) {
+    tf_shared_lock lock(mu_);
+    SHARED_LOCKS_REQUIRED(mu_) {
```
[the remark this answers is not in the document]
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the changes! Sorry for being ambiguous. I have clarified my requests about the thread safety attributes in the comments. The comments involving the outdated code don't need changes (I just felt like replying there too).

**Comment by bhavani-subramanian** [comment]
Thanks for the clarification, @penpornk. I have addressed your review comments.

**Review by penpornk — APPROVED** [review]
Thank you for the PR! :)

**Comment by bhavani-subramanian** [comment]
Hi @caisq the 2 failing checks are not related to this PR.

These checks fail for another recently submitted PR as well: https://github.com/tensorflow/tensorflow/pull/25186

**Comment by penpornk** [comment]
@bhavani-subramanian There seems to be a merge conflict. Could you please help take a look if you can resolve it from your side? Thank you!

**Comment by penpornk** [comment]
@bhavani-subramanian Never mind. It finally got pulled in now. Sorry!

**Comment by bhavani-subramanian** [comment]
Thanks @penpornk!

**Pull request merged**